akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -299,6 +437,239 @@ abstract class GraphStageLogic private[stream] (inCount: Int, outCount: Int) {
    */
   final protected def isClosed[T](out: Outlet[T]): Boolean = (interpreter.portStates(conn(out)) & OutClosed) != 0
 
+  private val emptyArray = new ArrayBuffer[Any](0)
+
+  /**
+   * Read a number of elements from the given inlet and continue with the given function,
+   * suspending execution if necessary. This action replaces the [[InHandler]]
+   * for the given inlet if suspension is needed and reinstalls the current
+   * handler upon receiving the last `onPush()` signal (before invoking the `andThen` function).
+   */
+  final protected def readN[T](in: Inlet[T], n: Int)(andThen: Seq[T] ⇒ Unit): Unit =
```

Comment:
do you think these methods (this and below) should be on the default GraphStageLogic? We can push them to a subclass, or a separate non-subclass that can be composed separately.